packages/google_maps_flutter/google_maps_flutter_web/lib/src/marker_clustering.dart

```diff
@@ -0,0 +1,205 @@
+  }
+
+  /// Adds given [gmaps.Marker] to the [MarkerClusterer] with given [ClusterManagerId].
+  void addItem(ClusterManagerId clusterManagerId, gmaps.Marker marker) {
+    final MarkerClusterer? markerClusterer =
+        _clusterManagerIdToMarkerClusterer[clusterManagerId];
+    if (markerClusterer != null) {
+      markerClusterer.addMarker(marker, true);
+      markerClusterer.dirty = true;
+    }
+  }
+
+  /// Removes given [gmaps.Marker] from the [MarkerClusterer] with given [ClusterManagerId].
+  void removeItem(ClusterManagerId clusterManagerId, gmaps.Marker? marker) {
+    if (marker != null) {
+      final MarkerClusterer? markerClusterer =
+          _clusterManagerIdToMarkerClusterer[clusterManagerId];
+      if (markerClusterer != null) {
+        markerClusterer.removeMarker(marker, true);
+        markerClusterer.dirty = true;
+      }
+    }
+  }
+
+  // Renders all cluster managers that are marked as dirty.
+  void renderDirty() {
+    _clusterManagerIdToMarkerClusterer.values
+        .where((MarkerClusterer markerClusterer) => markerClusterer.dirty)
+        .forEach((MarkerClusterer markerClusterer) {
+      markerClusterer.dirty = false;
+      markerClusterer.render();
+    });
+  }
+
+  /// Returns list of clusters in [MarkerClusterer] with given [ClusterManagerId].
+  List<Cluster> getClusters(ClusterManagerId clusterManagerId) {
+    final MarkerClusterer? markerClusterer =
+        _clusterManagerIdToMarkerClusterer[clusterManagerId];
+    if (markerClusterer != null) {
+      return markerClusterer.clusters
+          .map((MarkerClustererCluster cluster) =>
+              _convertCluster(clusterManagerId, cluster))
+          .toList();
+    }
+    return <Cluster>[];
+  }
+
+  void _clusterClicked(
+      ClusterManagerId clusterManagerId,
+      gmaps.MapMouseEvent event,
+      MarkerClustererCluster markerClustererCluster,
+      gmaps.GMap map) {
+    if (markerClustererCluster.count > 0 &&
+        markerClustererCluster.bounds != null &&
+        markerClustererCluster.markers != null) {
+      final Cluster cluster =
+          _convertCluster(clusterManagerId, markerClustererCluster);
+      _streamController.add(ClusterTapEvent(mapId, cluster));
+    }
+  }
+
+  /// Converts [MarkerClustererCluster] to [Cluster].
+  Cluster _convertCluster(ClusterManagerId clusterManagerId,
+      MarkerClustererCluster markerClustererCluster) {
+    final LatLng position = _gmLatLngToLatLng(markerClustererCluster.position);
+    final LatLngBounds bounds =
+        _gmLatLngBoundsTolatLngBounds(markerClustererCluster.bounds!);
+
+    final List<MarkerId> markerIds = markerClustererCluster.markers!
+        .map<MarkerId>((gmaps.Marker marker) =>
+            MarkerId(marker.get('markerId')! as String))
+        .toList();
+    return Cluster(
+      clusterManagerId,
+      markerIds,
+      position: position,
+      bounds: bounds,
+    );
+  }
+}
+
+@JS()
+external ClusterClickHandler defaultOnClusterClickHandler;
```

Comment:
I have a suspicion that these JS-interop bits need to be split to its own specific package (similar to `google_maps` or `google_identity_services_web`), actually the  new package should only have dependencies in `google_maps`, and then be used from `google_maps_flutter_web`, which would own converting between the JS-types and the ones from the `platform_interface`)

Reply:
Marker clustering is quite tightly mingled with the marker controller, so separating the whole markerclustering to separate package might be a bit tricky.
Are you suggesting that only the js interops are separated to separata package, or whole clustering support, as latter might be quite tricky to do.

Reply:
@jokerttu I meant only the `@JS` bits. It'd be a tiny package, but not baked into this package. If you split it to a separate file in this PR, we can move it to a separate package later.